lib/widgets/action_sheet.dart

```diff
@@ -372,37 +391,145 @@ class UserTopicUpdateButton extends ActionSheetMenuItemButton {
+
+      // The message's topic doesn't match what it was when we opened the
+      // action sheet. So either the resolve/unresolve action was already done,
+      // the topic was renamed, or at least one message in it including
+      // someMessageIdInTopic was moved. In any case, tell the user and don't
+      // make a request.
+      //
+      // This misses the case where some messages in the topic were moved
+      // but not someMessageIdInTopic. Shrug; discussion:
+      //   https://github.com/zulip/zulip-flutter/pull/1301#discussion_r1934565509
+      final title = _actionIsResolve
+        ? zulipLocalizations.resolveTopicInterruptedTitle
+        : zulipLocalizations.unresolveTopicInterruptedTitle;
+      final errorMessage = message.topic.unresolve().isSameAs(topic.unresolve())
+        ? _actionIsResolve
+          ? zulipLocalizations.topicAlreadyResolvedMessage
+          : zulipLocalizations.topicAlreadyUnresolvedMessage
+        // (Might not be a full topic rename; could have been a topic-move of
+        // just some messages. Shrug.)
+        : zulipLocalizations.topicRenamedMessage;
+      showErrorDialog(context: pageContext, title: title, message: errorMessage);
+      return;
+    }
+
+    if (message == null) {
+      // Proceed.
+      //
+      // This happens when [someMessageIdInTopic] came from [Unreads]
+      // and we just haven't fetched the message's details yet
+      // because we haven't opened a message list containing the message.
+      //
+      // Less commonly, this happens if the message was deleted
+      // after the action sheet was opened;
+      // see dartdoc of [ActionSheetMenuItemButton].
+      //
+      // Anyway, we have a message ID, which the server asks for; use it.
+    }
+
+    try {
+      await updateMessage(store.connection,
+        messageId: someMessageIdInTopic,
+        topic: _actionIsResolve ? topic.resolve() : topic.unresolve(),
+        propagateMode: PropagateMode.changeAll,
+        sendNotificationToOldThread: false,
+        sendNotificationToNewThread: true,
+      );
+    } catch (e) {
+      if (!pageContext.mounted) return;
+
+      String? errorMessage;
+
+      switch (e) {
+        case ZulipApiException():
+          errorMessage = e.message;
+          // TODO(#741) specific messages for common errors, like network errors
+          //   (support with reusable code)
+        default:
+      }
+
+      final title = _actionIsResolve
+        ? zulipLocalizations.errorResolveTopicFailedTitle
+        : zulipLocalizations.errorUnresolveTopicFailedTitle;
+      showErrorDialog(context: pageContext, title: title, message: errorMessage);
+    }
+  }
+}
+
 /// Show a sheet of actions you can take on a message in the message list.
 ///
-/// Must have a [MessageListPage] ancestor.
+/// Must have a [MessageListPage] ancestor and a [PageRoot] ancestor.
 void showMessageActionSheet({required BuildContext context, required Message message}) {
```

Comment:
I think specifically this implementation needs a PageRoot ancestor that *has* a MessageListPage ancestor.

But given that MessageListPage inserts a PageRoot, that just means a MessageListPage ancestor, right?